slimt/Annotation.cc

```diff
@@ -74,4 +77,81 @@ void AnnotatedText::record_existing_sentence(
   annotation.token_begin_.push_back(text.size());
 }
 
+void AnnotatedText::to(Encoding encoding) {
+  if (encoding == encoding_) return;
+  if (encoding == Encoding::Byte) {
+    WordIterator current(*this);
+
+    std::vector<Range> words;
+
+    size_t utf8_idx = 0;
+    size_t byte_idx = 0;
+    Range utf8{.begin = utf8_idx, .end = 0};
+    for (const char c : text) {
+      ++byte_idx;
+
+      if (byte_idx == (*current).end) {
+        utf8.end = utf8_idx;
+        words.push_back(utf8);
+        ++current;
+        utf8.begin = utf8_idx;
+      }
+    }
+    annotation.update(words);
+  } else if (encoding == Encoding::UTF8 && encoding_ == Encoding::Byte) {
+    WordIterator current(*this);
+
+    std::vector<Range> words;
+
+    size_t utf8_idx = 0;
+    size_t byte_idx = 0;
+    Range utf8{
+        //
+        .begin = utf8_idx,  //
+        .end = 0            //
+    };
+    // This loops run on the entire string.
+    //
+    // We have indices into two views of the same string. One is bytes, the
+    // other is utf8 encoded. We want to convert what is bytes to utf8.
+    //
+    // For this, we traverse through the characters, checking for unicode
+    // encoding and associated adjustments to the utf8_idx, keeping
+    // correspondences with the byte_idx;
+    for (const char c : text) {
+      // current = [begin, end)
+      if ((c & 0xc0) != 0x80)  // if is not utf-8 continuation character
+        ++utf8_idx;
+
+      ++byte_idx;
+
+      if (byte_idx == (*current).end) {
+        utf8.end = utf8_idx;
+
+        // Push it to the list of (utf8) words.
+        // We will use these to create the utf8 range annotation.
+        words.push_back(utf8);
+        ++current;
+
+        utf8.begin = utf8_idx;
+      }
+    }
+    annotation.update(words);
+  }
```

Comment:
`else { SLIMT_ABORT("Unimplemented"); }` will prevent us from tripping over when we have `UTF-16`.